### PR TITLE
Add new Provider interface for preview providers

### PR DESCRIPTION
**Pull request by icewind1991**
the main difference is passing the `File` object to the provider
instead of a `View` + path

Old providers will still continue to work as before, all builtin providers have been updated and received some cleanup in their temp/local file handling

The motivation for finally changing this is that the `View` + path doesn't provide enough flexibility to generate the correct previews for storage provided versions

### CODE REVIEW

File: lib/private/Preview/GeneratorHelper.php [hunk 1]
```diff
@@ -82,10 +70,14 @@ public function getImage(ISimpleFile $maxPreview) {
...
-	 * @return IProvider
+	 * @param $providerClosure
+	 * @return IProviderV2
 	 */
-	public function getProvider($provider) {
-		return $provider();
+	public function getProvider(callable $providerClosure) {
+		$provider = $providerClosure();
+		if ($provider instanceof IProvider) {
+			$provider = new ProviderV1Adapter($provider);
```
Comment (ChristophWurst) [inline lib/private/Preview/GeneratorHelper.php]: nice!

File: lib/public/Preview/IProviderV2.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @since 17.0.0
+ */
+interface IProviderV2 {
+	/**
+	 * @return string Regex with the mimetypes that are supported by this provider
+	 * @since 17.0.0
+	 */
+	public function getMimeType();
```
Comment (ChristophWurst) [inline lib/public/Preview/IProviderV2.php]: return type hint would be great :wink:

File: lib/public/Preview/IProviderV2.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+	/**
+	 * Check if a preview can be generated for $path
+	 *
+	 * @param FileInfo $file
+	 * @return bool
+	 * @since 17.0.0
+	 */
+	public function isAvailable(FileInfo $file);
```
Comment (ChristophWurst) [inline lib/public/Preview/IProviderV2.php]: same

File: lib/public/Preview/IProviderV2.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+	/**
+	 * get thumbnail for file at path $path
+	 *
+	 * @param File $file
+	 * @param int $maxX The maximum X size of the thumbnail. It can be smaller depending on the shape of the image
+	 * @param int $maxY The maximum Y size of the thumbnail. It can be smaller depending on the shape of the image
+	 * @return bool|\OCP\IImage false if no preview was generated
```
Comment (ChristophWurst) [inline lib/public/Preview/IProviderV2.php]: how about returning `null` instead of `false`? Then a return type could be specified

**Review by ChristophWurst — COMMENTED** [review]
Looks good!

File: lib/private/Preview/ProviderV2.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	abstract public function getThumbnail(File $file, int $maxX, int $maxY);
+
+	/**
+	 * Get a path to either the local file or temporary file
+	 *
+	 * @param File $file
+	 * @param int $maxSize maximum size for temporary files
+	 * @return string
+	 */
+	protected function getLocalFile(File $file, int $maxSize = null): string {
```
Comment (juliushaertl) [inline lib/private/Preview/ProviderV2.php]: :tada:

File: lib/private/Preview/SVG.php [hunk 6]
```diff
@@ -37,12 +38,12 @@ public function getMimeType() {
...
 	 * {@inheritDoc}
 	 */
-	public function getThumbnail($path, $maxX, $maxY, $scalingup, $fileview) {
+	public function getThumbnail(File $file, int $maxX, int $maxY) {
 		try {
 			$svg = new \Imagick();
 			$svg->setBackgroundColor(new \ImagickPixel('transparent'));
 
-			$content = stream_get_contents($fileview->fopen($path, 'r'));
+			$content = stream_get_contents($file->fopen('r'));
```
Comment (juliushaertl) [inline lib/private/Preview/SVG.php]: This should probably also use $this->getLocalFile()

File: lib/private/Preview/ProviderV2.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+		$useTempFile = $file->isEncrypted() || !$file->getStorage()->isLocal();
+		if ($useTempFile) {
+			$absPath = \OC::$server->getTempManager()->getTemporaryFile();
+
+			$content = $file->fopen('r');
+
+			if ($maxSize) {
+				$content = stream_get_contents($content, 5242880);
```
Comment (juliushaertl) [inline lib/private/Preview/ProviderV2.php]: Should be $maxSize instead of the `5242880`

**Review by juliusknorr — CHANGES_REQUESTED** [review]

File: lib/private/Preview/SVG.php (hunk 6 above)
Comment (icewind1991) [reply lib/private/Preview/SVG.php]: shouldn't make a difference here, local files are only useful if the generator require a local path to work with

File: lib/private/Preview/ProviderV2.php (hunk 7 above)
Comment (icewind1991) [reply lib/private/Preview/ProviderV2.php]: fixed

**Comment by icewind1991** [comment]
added type hints

**Review by ChristophWurst — APPROVED** [review]

**Review by rullzer — APPROVED** [review]
Lets do this!

**Comment by skjnldsv** [comment]
It broke master!
Let me fix :)

**Pull request merged**